contracts/upgradeable_contracts/arbitrary_message/BasicHomeAMB.sol

```diff
@@ -0,0 +1,137 @@
+pragma solidity 0.4.24;
+
+import "../../libraries/Message.sol";
+import "./BasicAMB.sol";
+import "./MessageDelivery.sol";
+import "./MessageProcessor.sol";
+
+contract BasicHomeAMB is BasicAMB, MessageDelivery, MessageProcessor {
+    event SignedForUserRequest(address indexed signer, bytes32 messageHash);
+    event SignedForAffirmation(address indexed signer, bytes32 messageHash);
+
+    event CollectedSignatures(
+        address authorityResponsibleForRelay,
+        bytes32 messageHash,
+        uint256 NumberOfCollectedSignatures
+    );
+
+    function executeAffirmation(bytes message) external onlyValidator {
+        bytes32 hashMsg = keccak256(abi.encodePacked(message));
+        bytes32 hashSender = keccak256(abi.encodePacked(msg.sender, hashMsg));
+        // Duplicated affirmations
+        require(!affirmationsSigned(hashSender));
+        setAffirmationsSigned(hashSender, true);
+
+        uint256 signed = numAffirmationsSigned(hashMsg);
+        require(!isAlreadyProcessed(signed));
+        // the check above assumes that the case when the value could be overflew will not happen in the addition operation below
+        signed = signed + 1;
+
+        setNumAffirmationsSigned(hashMsg, signed);
+
+        emit SignedForAffirmation(msg.sender, hashMsg);
+
+        if (signed >= requiredSignatures()) {
+            setNumAffirmationsSigned(hashMsg, markAsProcessed(signed));
+            processMessage(message, false);
+        }
+    }
+
+    function submitSignature(bytes signature, bytes message) external onlyValidator {
+        // ensure that `signature` is really `message` signed by `msg.sender`
+        require(msg.sender == Message.recoverAddressFromSignedMessage(signature, message, true));
+        bytes32 hashMsg = keccak256(abi.encodePacked(message));
+        bytes32 hashSender = keccak256(abi.encodePacked(msg.sender, hashMsg));
+
+        uint256 signed = numMessagesSigned(hashMsg);
+        require(!isAlreadyProcessed(signed));
+        // the check above assumes that the case when the value could be overflew
+        // will not happen in the addition operation below
+        signed = signed.add(1);
```

Comment:
We don't use `safeMath` in the similar handling within `executeAffirmation`. Let's keep it consistent.

Reply:
Fixed here https://github.com/poanetwork/poa-bridge-contracts/pull/77/commits/1613a10420cbe938a23a38847a4235aec0af494b